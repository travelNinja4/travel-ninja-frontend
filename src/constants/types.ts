/**
 * used to store global types in the project
 *
 */

export type OtpProps = {
  otpCode: string;
  otpType?: 'email' | 'mobile';
};

export interface LoginResponse {
  data: {
    phoneNumber?: string;
  };
}
