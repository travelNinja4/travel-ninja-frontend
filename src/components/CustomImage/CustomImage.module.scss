/**
------------------------------------------------------
styles specific to the CustomImage component
------------------------------------------------------
**/

.container {
  padding: 1rem;
  // Add styles here
}
