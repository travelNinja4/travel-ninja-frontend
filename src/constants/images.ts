/**
 * Handles all the cloudinary image constants
 *
 */

export const IMAGES = {
  FALLBACK_IMG: 'fallbackImg.png',
};
