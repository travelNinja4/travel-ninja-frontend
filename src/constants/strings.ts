export const STRINGS = {
  APP_NAME: 'TravelNinja',
  CREATE_ACCOUNT_APP_TAG_LINE: 'Transform your tour business with our powerful platform',
  LOGIN_APP_TAG_LINE: 'Empowering tour agencies to create unforgettable travel experiences',
  RESET_PASSWORD_TAG_LINE: 'Reset your password to regain access to your account',
  EMAIL: 'email',
  MOBILE: 'mobile',
  EMAIL_VERIFICATION: 'Email Verification',
  MOBILE_NUMBER_VERIFICATION: 'Mobile Number Verification',
  EMAIL_SENT: `We've sent a verification code to your email`,
  MOBILE_SENT: `We've sent a verification code to your mobile number`,
  EMAIL_CODE_HINT: 'Enter the 6-digit code sent to your email',
  MOBILE_CODE_HINT: 'Enter the 6-digit code sent to your mobile number',
  VERIFY_EMAIL: 'Verify Email',
  VERIFY_MOBILE_NUMBER: 'Verify Mobile Number',
  ENTER_VERIFICATION_CODE: 'Enter Verification Code',
  RESEND_CODE_IN: 'Resend code in',
  SECONDS: 'seconds',
  RESEND_CODE: 'Resend Code',
  NO_OPTIONS_FOUND: 'No options found',
  COUNTRY_CODE: 'Country code',
  ENTER_MOBILE_NUMBER: 'Enter mobile number',
  FORGOT_PASSWORD: 'Forgot Password?',
  RESET_PASSWORD_INSTRUCTION: 'Enter your email to receive reset instructions',
  RESET_PASSWORD_LABEL: `We'll send you a link to reset your password`,
  SEND_RESET_LINK: 'Send Reset Link',
  EMAIL_ADDRESS: 'Email Address',
  ENTER_EMAIL_ADDRESS: 'Enter your email address',
  BACK_TO_LOGIN: 'Back to Login',
};

export const ROUTES = {
  HOME: '/',
  LOGIN: '/login',
  CREATE_ACCOUNT: '/createAccount',
  VERIFY: '/verify',
} as const;

export const NOTIFICATION_TYPES = {
  SUCCESS: 'success',
  ERROR: 'error',
  WARNING: 'warning',
  INFO: 'info',
};
