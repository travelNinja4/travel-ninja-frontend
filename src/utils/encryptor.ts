import crypto from 'crypto';

const publicKey = process.env.NEXT_PUBLIC_RSA_PUBLIC_KEY as string;
const privateKey = process.env.RSA_PRIVATE_KEY as string;

/**
 * Encrypts a string value using RSA public key encryption (OAEP with SHA-256).
 * @param {string} value - The plaintext string to encrypt.
 * @returns {string} The encrypted value, encoded in base64.
 */
function encryptValue(value: string): string {
  const buffer = Buffer.from(value, 'utf8');
  const encrypted = crypto.publicEncrypt(
    {
      key: publicKey,
      padding: crypto.constants.RSA_PKCS1_OAEP_PADDING,
      oaepHash: 'sha256',
    },
    buffer,
  );
  return encrypted.toString('base64');
}

/**
 * Decrypts a string value using RSA private key decryption (OAEP with SHA-256).
 *
 * @param {string} encryptedValue - The encrypted value in base64 format.
 * @returns {string} The decrypted plaintext string.
 */
function decryptValue(encryptedValue: string): string {
  const buffer = Buffer.from(encryptedValue, 'base64');
  const decrypted = crypto.privateDecrypt(
    {
      key: privateKey,
      padding: crypto.constants.RSA_PKCS1_OAEP_PADDING,
      oaepHash: 'sha256',
    },
    buffer,
  );
  return decrypted.toString('utf8');
}

/**
 * Encrypts specific string fields in an object using RSA public key encryption.
 *
 * @param {T} data - The object containing fields to encrypt.
 * @param {string[]} keysToEncrypt - The keys of the object that should be encrypted.
 * @returns {T} A new object with the specified fields encrypted.
 */
export function encryptObject<T extends Record<string, any>>(data: T, keysToEncrypt: string[]): T {
  const result: Record<string, any> = { ...data };

  keysToEncrypt.forEach((key) => {
    if (result[key] && typeof result[key] === 'string') {
      result[key] = encryptValue(result[key]);
    }
  });

  return result as T;
}

/**
 * Decrypts specific string fields in an object using RSA private key decryption.
 *
 * @param {T} data - The object containing fields to decrypt.
 * @param {string[]} keysToDecrypt - The keys of the object that should be decrypted.
 * @returns {T} A new object with the specified fields decrypted.
 */
export function decryptObject<T extends Record<string, any>>(data: T, keysToDecrypt: string[]): T {
  const result: Record<string, any> = { ...data };

  keysToDecrypt.forEach((key) => {
    if (result[key] && typeof result[key] === 'string') {
      result[key] = decryptValue(result[key]);
    }
  });

  return result as T;
}
