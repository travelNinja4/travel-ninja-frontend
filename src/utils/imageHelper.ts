/**
 * Provides helper functions for working with images
 *
 */

export const buildCloudinaryUrl = (imageName: string) =>
  `https://res.cloudinary.com/dyci0wg0i/image/upload/${imageName}`;
