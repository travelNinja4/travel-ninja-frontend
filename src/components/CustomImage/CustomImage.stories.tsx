import type { Meta, StoryObj } from '@storybook/nextjs-vite';
/**
import { fn } from 'storybook/test';
*/
import CustomImage from './CustomImage';

const meta = {
  title: 'components/CustomImage',
  component: CustomImage,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    src: {
      control: 'text',
      description: 'Image source key from IMAGES or a direct URL',
    },
    alt: {
      control: 'text',
      description: 'Accessible alt text for the image',
    },
    width: {
      control: 'number',
      description: 'Image width (ignored if fill=true)',
    },
    height: {
      control: 'number',
      description: 'Image height (ignored if fill=true)',
    },
    fill: {
      control: 'boolean',
      description: 'If true, the image will fill its container',
    },
    isUrl: {
      control: 'boolean',
      description: 'If true, src is treated as a direct URL instead of IMAGES key',
    },
    loading: {
      control: { type: 'radio' },
      options: ['lazy', 'eager'],
      description: 'Image loading strategy',
    },
    priority: {
      control: 'boolean',
      description: 'If true, loads image with high priority',
    },
  },
  args: {
    src: 'FALLBACK_IMG',
    alt: 'Default Image',
    width: 200,
    height: 200,
    fill: false,
    isUrl: false,
    loading: 'lazy',
    priority: false,
  },
} satisfies Meta<typeof CustomImage>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {};

export const WithFallback: Story = {
  args: {
    src: 'INVALID_KEY',
    alt: 'Fallback image demo',
  },
};

export const CloudinaryImage: Story = {
  args: {
    src: 'FALLBACK_IMG',
    alt: 'Cloudinary image example',
    width: 300,
    height: 200,
  },
};

export const RemoteUrlImage: Story = {
  args: {
    src: 'https://picsum.photos/300/200',
    alt: 'External URL image',
    isUrl: true,
  },
};

export const FillContainer: Story = {
  args: {
    src: 'FALLBACK_IMG',
    alt: 'Fill container demo',
    fill: true,
    style: { objectFit: 'cover' },
  },
};
