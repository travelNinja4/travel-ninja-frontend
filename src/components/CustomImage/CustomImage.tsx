/**
 * A flexible, reusable image component built on top of Next.js Image, with built-in fallback, responsive handling, and accessibility checks to ensure optimized and consistent image rendering across the application.
 *
 * @example
 * ```tsx
 * import CustomImage from '@src/components/CustomImage'
 *
 * export default function CustomImage() {
 *   return <CustomImage
 *             src="https://picsum.photos/400/250"
 *             alt="Random placeholder"
 *             isUrl width={400}
 *             height={250}
 *       />;
 *  }
 * ```
 */
'use client';

import Image from 'next/image';
import { IMAGES } from '@/constants/images';
import type { ImageProps } from 'next/image';
import { buildCloudinaryUrl } from '@/utils/imageHelper';
import styles from './CustomImage.module.scss';

/**
 * Define the props available for the CustomImage component.
 */
interface CustomImageProps extends Omit<ImageProps, 'src' | 'alt'> {
  /**
   * The source of the image.
   *
   * If `isUrl` is `true`, this should be a direct image URL.
   * If `isUrl` is `false`, this should be a key from the `IMAGES` constant.
   * If empty/invalid, the fallback image will be used.
   */
  src: string;

  /**
   *
   */
  alt: string;

  /**
   * The width of the image (in pixels).
   * Ignored if `fill` is set to `true`.
   *
   * @default 200
   */
  width?: number;

  /**
   * The height of the image (in pixels).
   * Ignored if `fill` is set to `true`.
   *
   * @default 200
   */
  height?: number;

  /**
   * If `true`, the image will stretch to fill its parent container
   * while maintaining aspect ratio.
   *
   * @default false
   */
  fill?: boolean;

  /**
   * Inline styles applied to the image container.
   */
  style?: object;

  /**
   * Optional CSS class name for applying custom styles.
   */
  className?: string;

  /**
   * Defines the loading behavior of the image.
   *
   * `lazy`: Image loads only when visible on the viewport (default).
   * `eager`: Image loads immediately, regardless of visibility.
   *
   * @default 'lazy'
   */
  loading?: 'lazy' | 'eager';

  /**
   * If `true`, the image is considered high priority and will preload.
   * Useful for above-the-fold images like banners or logos.
   *
   * @default false
   */
  priority?: boolean;

  /**
   * Whether the `src` provided is a full URL or a key from `IMAGES`.
   *
   * `true`: Treats `src` as a valid URL.
   * `false`: Maps `src` to an entry in the `IMAGES` constant.
   *
   * @default false
   */
  isUrl?: boolean;
}

export default function CustomImage({
  src,
  alt,
  width = 200,
  height = 200,
  fill = false,
  style = {},
  className,
  loading = 'lazy',
  priority = false,
  isUrl = false,
  ...rest
}: CustomImageProps) {
  const resolvedSrc = (() => {
    if (!src) return buildCloudinaryUrl(IMAGES.FALLBACK_IMG);
    if (isUrl) return src;
    const mappedSrc = IMAGES[src as keyof typeof IMAGES];
    return mappedSrc ? buildCloudinaryUrl(mappedSrc) : buildCloudinaryUrl(IMAGES.FALLBACK_IMG);
  })();

  return (
    <Image
      data-testid="CustomImageTest"
      src={resolvedSrc}
      alt={alt}
      {...(!fill && { width, height })}
      {...(fill && { fill })}
      style={style}
      className={className}
      {...(!priority && { loading })}
      priority={priority}
      placeholder="blur"
      blurDataURL={buildCloudinaryUrl(IMAGES.FALLBACK_IMG)}
      onError={(e) => {
        const target = e.target as HTMLImageElement;
        target.src = buildCloudinaryUrl(IMAGES.FALLBACK_IMG);
      }}
      {...rest}
    />
  );
}
