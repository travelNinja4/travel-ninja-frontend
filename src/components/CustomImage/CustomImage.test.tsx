import '@testing-library/jest-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import { buildCloudinaryUrl } from '@/utils/imageHelper';
import { IMAGES } from '@/constants/images';
import CustomImage from './CustomImage';

describe('Home', () => {
  /** Base props for CustomImage **/
  const baseProps = {
    src: 'FALLBACK_IMG',
    alt: 'CustomImage Test',
  };

  /**
   * @test
   * Renders the CustomImage component without throwing an error
   * and checks if the CustomImageTest text is present in the document.
   *
   * This ensures that the component loads correctly with the base props.
   */
  it('renders without crashing', () => {
    render(<CustomImage {...baseProps} />);
    expect(screen.getByTestId('CustomImageTest')).toBeInTheDocument();
  });

  /**
   * @test
   * Ensures that the CustomImage component renders without throwing any runtime errors.
   *
   * This is a safety check to confirm that the component can mount successfully
   * with the provided base props, even if it does not render any visible content.
   */
  it('renders without throwing', () => {
    expect(() => render(<CustomImage {...baseProps} />)).not.toThrow();
  });

  /**
   * @test
   * Captures a snapshot of the CustomImage component with base props
   * and ensures it matches the previously stored snapshot.
   *
   * This helps detect any unexpected UI changes.
   */
  it('matches snapshot', () => {
    const { asFragment } = render(<CustomImage {...baseProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('uses fallback image when src is empty', () => {
    render(<CustomImage {...baseProps} src="" />);
    const img = screen.getByTestId('CustomImageTest') as HTMLImageElement;
    expect(img).toHaveAttribute('src', buildCloudinaryUrl(IMAGES.FALLBACK_IMG));
  });

  it('renders direct URL when isUrl is true', () => {
    const url = 'https://example.com/test.jpg';
    render(<CustomImage {...baseProps} src={url} isUrl />);
    const img = screen.getByTestId('CustomImageTest') as HTMLImageElement;
    expect(img).toHaveAttribute('src', url);
  });

  it('uses fallback when src key is not in IMAGES', () => {
    render(<CustomImage {...baseProps} src="INVALID_KEY" />);
    const img = screen.getByTestId('CustomImageTest') as HTMLImageElement;
    expect(img).toHaveAttribute('src', buildCloudinaryUrl(IMAGES.FALLBACK_IMG));
  });

  it('switches to fallback image on error', () => {
    render(<CustomImage {...baseProps} isUrl src="https://broken-url.com/image.png" />);
    const img = screen.getByTestId('CustomImageTest') as HTMLImageElement;
    fireEvent.error(img);
    expect(img.src).toBe(buildCloudinaryUrl(IMAGES.FALLBACK_IMG));
  });
});
