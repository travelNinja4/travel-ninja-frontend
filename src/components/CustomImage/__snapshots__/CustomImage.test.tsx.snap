// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Home matches snapshot 1`] = `
<DocumentFragment>
  <img
    alt="CustomImage Test"
    blurdataurl="https://res.cloudinary.com/dyci0wg0i/image/upload/fallbackImg.png"
    data-testid="CustomImageTest"
    height="200"
    loading="lazy"
    placeholder="blur"
    src="https://res.cloudinary.com/dyci0wg0i/image/upload/fallbackImg.png"
    width="200"
  />
</DocumentFragment>
`;
